src/v3/plugins/github/graphView.js

```diff
@@ -0,0 +1,274 @@
+    this._maybeCheckInvariants();
+  }
+
+  graph(): Graph {
+    this._maybeCheckInvariants();
+    return this._graph;
+  }
+
+  *_nodes<T: GN.StructuredAddress>(prefix: GN.RawAddress): Iterator<T> {
+    for (const n of this._graph.nodes({prefix})) {
+      const structured = GN.fromRaw((n: any));
+      this._maybeCheckInvariants();
+      yield (structured: any);
+    }
+    this._maybeCheckInvariants();
+  }
+
+  *_neighbors<T: GN.StructuredAddress>(
+    node: GN.StructuredAddress,
+    options: NeighborsOptions
+  ): Iterator<T> {
+    if (!NodeAddress.hasPrefix(options.nodePrefix, GN._Prefix.base)) {
+      throw new Error(`_neighbors must filter to Github nodes`);
+    }
+    const rawNode = GN.toRaw(node);
+    for (const neighbor of this._graph.neighbors(rawNode, options)) {
+      this._maybeCheckInvariants();
+      yield (GN.fromRaw((neighbor.node: any)): any);
+    }
+    this._maybeCheckInvariants();
+  }
+
+  _children<T: GN.StructuredAddress>(
+    node: GN.StructuredAddress,
+    nodePrefix: GN.RawAddress
+  ): Iterator<T> {
+    const options = {
+      nodePrefix,
+      edgePrefix: GE._Prefix.hasParent,
+      direction: Direction.IN,
+    };
+    return this._neighbors(node, options);
+  }
+
+  repos(): Iterator<GN.RepoAddress> {
+    return this._nodes(GN._Prefix.repo);
+  }
+
+  issues(repo: GN.RepoAddress): Iterator<GN.IssueAddress> {
+    return this._children(repo, GN._Prefix.issue);
+  }
+
+  pulls(repo: GN.RepoAddress): Iterator<GN.PullAddress> {
+    return this._children(repo, GN._Prefix.pull);
+  }
+
+  comments(commentable: GN.CommentableAddress): Iterator<GN.CommentAddress> {
+    return this._children(commentable, GN._Prefix.comment);
+  }
+
+  reviews(pull: GN.PullAddress): Iterator<GN.ReviewAddress> {
+    return this._children(pull, GN._Prefix.review);
+  }
+
+  // TODO figure out how to overload this fn signature
+  parent(child: GN.ChildAddress): GN.ParentAddress {
+    const options = {
+      direction: Direction.OUT,
+      edgePrefix: GE._Prefix.hasParent,
+      nodePrefix: GN._Prefix.base,
+    };
+    const parents: GN.ParentAddress[] = Array.from(
+      this._neighbors(child, options)
+    );
+    if (parents.length !== 1) {
+      throw new Error(
+        `Parent invariant violated for child: ${stringify(child)}`
+      );
+    }
+    return parents[0];
+  }
+
+  authors(content: GN.AuthorableAddress): Iterator<GN.UserlikeAddress> {
+    const options = {
+      direction: Direction.IN,
+      edgePrefix: GE._Prefix.authors,
+      nodePrefix: GN._Prefix.userlike,
+    };
+    return this._neighbors(content, options);
+  }
+
+  _maybeCheckInvariants() {
+    if (this._isCheckingInvariants) {
+      return;
+    }
+    if (process.env.NODE_ENV === "test") {
+      // TODO(perf): If this method becomes really slow, we can disable
+      // it on specific tests wherein we construct large graphs.
+      this.checkInvariants();
+    }
+  }
+
+  checkInvariants() {
```

Comment:
I would like to see a test that this is properly reentrant, perhaps like:

```js
const g = exampleGraph();
const view = new GraphView(g);
// Deliberately corrupt the invariants: add an extra `hasParent` edge or
// something.
try {
  view.graph();
} catch (_) {}
expect(() => {
  view.graph();
}).toThrow("Invariant violation");
```

(such that the test fails without the `finally`).

Reply:
Done.